# Semana 9 - Ejercicio Integrador: Riesgo Crediticio

**Materia:** Minería de Datos  
**Tema:** Árboles de Decisión y Random Forest  
**Estudiante:** Sara Valenzuela

## Objetivo

Construir modelos de clasificación para predecir el riesgo de incumplimiento crediticio y comparar el desempeño de un Árbol de Decisión con Random Forest.

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import (
    train_test_split,
    GridSearchCV
)

from sklearn.tree import DecisionTreeClassifier

from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import (
    accuracy_score,
    classification_report
)

In [2]:
# Generar dataset sintético

np.random.seed(42)

data = {
    "Edad": np.random.randint(18, 70, 500),
    "Ingresos": np.random.randint(20000, 150000, 500),
    "Historial": np.random.randint(1, 10, 500),
    "Monto_Credito": np.random.randint(5000, 100000, 500),
    "Incumplimiento": np.random.choice(
        [0, 1],
        500,
        p=[0.7, 0.3]
    )
}

df = pd.DataFrame(data)

df.head()

,Edad,Ingresos,Historial,Monto_Credito,Incumplimiento
0,56,23343,1,93929,0
1,69,33500,5,28051,1
2,46,73222,4,56474,1
3,32,49375,8,96748,0
4,60,29662,1,74626,0


## Descripción del Dataset

El conjunto de datos representa información simulada de solicitantes de crédito.

La variable objetivo es **Incumplimiento**, donde:

- 0 = No incumple
- 1 = Incumple

El objetivo es predecir el riesgo crediticio a partir de características demográficas y financieras.

In [3]:
X = df.drop(
    columns=["Incumplimiento"]
)

y = df["Incumplimiento"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

print(X_train.shape)
print(X_test.shape)

(400, 4)
(100, 4)


In [4]:
arbol = DecisionTreeClassifier(
    random_state=42
)

arbol.fit(
    X_train,
    y_train
)

pred_arbol = arbol.predict(
    X_test
)

acc_arbol = accuracy_score(
    y_test,
    pred_arbol
)

print("Accuracy Árbol:", acc_arbol)

Accuracy Árbol: 0.62


In [5]:
param_grid = {
    "max_depth": [2, 3, 4, 5, 6],
    "criterion": ["gini", "entropy"]
}

grid = GridSearchCV(
    DecisionTreeClassifier(random_state=42),
    param_grid,
    cv=5,
    scoring="accuracy"
)

grid.fit(
    X_train,
    y_train
)

print(grid.best_params_)

{'criterion': 'gini', 'max_depth': 4}


In [6]:
arbol_opt = grid.best_estimator_

pred_opt = arbol_opt.predict(
    X_test
)

acc_opt = accuracy_score(
    y_test,
    pred_opt
)

print("Accuracy Árbol Optimizado:", acc_opt)

Accuracy Árbol Optimizado: 0.76


In [7]:
rf = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf.fit(
    X_train,
    y_train
)

pred_rf = rf.predict(
    X_test
)

acc_rf = accuracy_score(
    y_test,
    pred_rf
)

print("Accuracy Random Forest:", acc_rf)

Accuracy Random Forest: 0.67


In [8]:
comparacion = pd.DataFrame({
    "Modelo": [
        "Árbol",
        "Árbol Optimizado",
        "Random Forest"
    ],
    "Accuracy": [
        acc_arbol,
        acc_opt,
        acc_rf
    ]
})

comparacion

,Modelo,Accuracy
0,Árbol,0.62
1,Árbol Optimizado,0.76
2,Random Forest,0.67


In [9]:
print(
    classification_report(
        y_test,
        pred_rf
    )
)

              precision    recall  f1-score   support

           0       0.75      0.84      0.79        75
           1       0.25      0.16      0.20        25

    accuracy                           0.67       100
   macro avg       0.50      0.50      0.49       100
weighted avg       0.62      0.67      0.64       100



## Conclusión Final

En este ejercicio integrador se compararon diferentes modelos de clasificación aplicados al problema de riesgo crediticio.

El árbol de decisión básico obtuvo una precisión de 62%, mientras que la optimización mediante GridSearchCV permitió incrementar el rendimiento hasta 76%, identificando como mejor configuración un árbol con criterio Gini y profundidad máxima de 4 niveles.

Posteriormente se entrenó un modelo Random Forest, alcanzando una precisión de 67%. Aunque Random Forest suele ofrecer un rendimiento superior en muchos problemas de clasificación, en este conjunto de datos el árbol optimizado obtuvo mejores resultados.

Los resultados muestran la importancia de la selección y optimización de hiperparámetros para mejorar la capacidad predictiva de los modelos. Asimismo, permiten comparar distintas técnicas de aprendizaje supervisado utilizadas habitualmente en problemas de análisis de riesgo crediticio.

En conclusión, el árbol de decisión optimizado fue el modelo con mejor desempeño para este conjunto de datos, alcanzando una precisión del 76%.